### ISSUE TITLE
Connection to cluster

### ISSUE THREAD

**Issue by toabi** [Shogan/kube-chaos#1]
I have a working kubeconfig, when I do `kubectl get -n somespace all` I see things… but when I launch the game and point to that namespace I see nothing. It launches into an empty "space" :D 

Are there any special things to do before connecting?

**Comment by Shogan** [comment]
Hi @toabi,

The code is effectively launching a process with kubectl and passing the parameters for your namespace and selected context through to that.

What platform are you running on if you're running from source? Or are you using the pre-compiled 1.0.0 release?

Could you also double check that the kubectl is on your system path? I think the .NET Process launch expects that, as I don't pass in an absolute path to kubectl, instead relying on it being in the system path.

**Comment by Shogan** [comment]
Hey @toabi ,

Just following up again on this. Did you check to see if kubectl is on your system path?

Another thing you could try is to issue the exact command that the game issues when building out the node objects in the starting scene (where you would see them floating around and are able to 'enter' them).

You could make sure that this also returns a simple named list of your nodes:

`kubectl --context {your-kube-context-name-here} get nodes -o custom-columns=:metadata.name`

If that returns your nodes list, and kubectl is on your system path then it should be working. Otherwise it'll also be useful to know your OS and if you're running the 1.0.0 release binary, or if you are running direct from source using the Unity editor :+1:

Thanks!

**Comment by toabi** [comment]
A colleague built it for me because he already had a unity build setup.

I also tried it with full-path to kubectl (although it's in $PATH) and the context which works in my shell… but in the game I don't see anything still.

```
❯ kubectl --context "testing" get pod -n myspace
NAME                           READY   STATUS    RESTARTS   AGE
celery-7c6fd75954-2g7nn        1/1     Running   0          30d
celery-beat-68d6949f7f-hsmht   1/1     Running   0          30d
django-57c5c87fb9-f4gfz        1/1     Running   0          30d
django-57c5c87fb9-jvl8z        1/1     Running   0          30d
django-57c5c87fb9-wt28c        1/1     Running   0          30d
manual-85945f7fcc-6jdqn        1/1     Running   0          113d
memcached-755476bfd8-6hv6h     1/1     Running   0          108d
rabbitmq-bbb7565f7-78948       1/1     Running   0          108d
```

So in the UI I put "myspace" / "testing" / "kubectl"

**Comment by Shogan** [comment]
@toabi I'm a little stumped then, unless something changed in the build your colleague did.

What platform/OS are you on? If you're on Windows, I would suggest trying the current 1.0.0 release version (pre-built), otherwise if you're macOs or Linux I can put together a release with binary for that OS.

Just let me know and I'll get it built for you to try.

Thanks!

**Comment by toabi** [comment]
I am on macOS. A build would be great.

**Comment by Shogan** [comment]
> I am on macOS. A build would be great.

Thanks for that @toabi.

I loaded it up on my macOS machine this evening in the editor to get logging, tried it out and saw an issue immediately. The default `kubectl` isn't recognised in the system path.

I know you said you tried putting in the full path to kubectl on your machine, but could you double check that is right? For me:

```
~/Git which kubectl                                                                                                                                                                       
/usr/local/bin/kubectl
```

So after entering `/usr/local/bin/kubectl` as the kubectl executable on start, the game works fine for me now :+1:

TLDR;

If you're on macOS, definitely provide the absolute path to `kubectl` on your system.

e.g. on my macOS system:

<img width="313" alt="start-macOs" src="https://user-images.githubusercontent.com/814180/89473342-8674b980-d77a-11ea-9bab-03f9dd28fb4b.png">
<img width="1638" alt="screen1" src="https://user-images.githubusercontent.com/814180/89473349-88d71380-d77a-11ea-86a8-903f3375cd65.png">

I've also built a macOS release. I'm attaching it shortly.

**Comment by Shogan** [comment]
One last thing @toabi , I ran the compiled build and even after entering the absolute path to kubectl on my system, I saw the issue you had. It seems to be timing related. You definitely need the absolute path to kubectl to be entered on start in the main menu, but if that doesn't work the first couple of times, try again. On my second try it worked.

I think it's related to the initial thread I create to execute the kubectl process to fetch the initial nodes list. A simple hacky fix might be to just delay that call by a second or two on loading that initial scene.

**Comment by toabi** [comment]
Thanks for the build. Indeed the cluster I was using for testing has a quite slow connection and it can take long (multiple seconds) to receive node info.

But I now also tried it with a EKS cluster where the answer for nodes is usally around 700ms … still no success :(

**Comment by Shogan** [comment]
> Thanks for the build. Indeed the cluster I was using for testing has a quite slow connection and it can take long (multiple seconds) to receive node info.
> 
> But I now also tried it with a EKS cluster where the answer for nodes is usally around 700ms … still no success :(

OK, I think I have your problem fixed now. It's a little hacky and just waits 1.5 seconds before loading the main scene to give the node info fetching thread time to return the list of nodes back.

Try the attached build out, and let me know :)

If it works I'll push changes and publish that as a 1.0.1 release.

[KubeChaos_macOS_1.0.1.zip](https://github.com/Shogan/kube-chaos/files/5034348/KubeChaos_macOS_1.0.1.zip)

**Comment by toabi** [comment]
Weird, when I copy-pasted it didn't work, when I started writing out namespace/context it did.

When flying into a node it gets stuck for quite some time and then continues but I don't see a pod :( although there should be at least node-exporter :D 

Nodes also just show "Node: loading…" for ever. (Ah, no it just takes really long, I don't know why, with kubectl it's fast)

**Comment by toabi** [comment]
Small success! <img width="711" alt="image" src="https://user-images.githubusercontent.com/115045/89524895-a2fa0b80-d7e5-11ea-86e5-d8bebfa58cfc.png">

**Comment by toabi** [comment]
Okay after flying into the one with filled info I now see pods!

**Comment by Shogan** [comment]
> Nodes also just show "Node: loading…" for ever. (Ah, no it just takes really long, I don't know why, with kubectl it's fast)

node updates are on a timer, every 8-10 seconds they get an update from the background thread, so that might be why. But you should definitely see them showing the correct info after 10 seconds or so, if the initial info population did not work.

> Okay after flying into the one with filled info I now see pods!

Great! Glad it's working. 

It's not perfect with the timed updates. Ideally I would be using events from the kube API for this stuff, but unfortunately the C# kubernetes-client library doesn't play nicely with Unity (or rather Unity does not play nicely with NuGet).

A nicer option would be to either run a small proxy API alongside that does use the C# kubernetes-client library, or to get that library working in the game's codebase :)

Thanks for testing it out. We're at least in a better state for others now that might have a bit of a delay / remote kube clusters. :+1:

I never saw the issue myself as I always tested with a fairly local kube cluster (my rPi cluster which is only a small wifi hop away).
